app/src/main/java/org/mozilla/focus/telemetry/GleanMetricsService.kt

```diff
@@ -0,0 +1,103 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.telemetry
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.search.SearchEngine
+import mozilla.components.browser.state.state.selectedOrDefaultSearchEngine
+import mozilla.components.feature.search.ext.waitForSelectedOrDefaultSearchEngine
+import mozilla.components.service.glean.net.ConceptFetchHttpUploader
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.config.Configuration
+import org.mozilla.focus.BuildConfig
+import org.mozilla.focus.Components
+import org.mozilla.focus.GleanMetrics.GleanBuildInfo
+import org.mozilla.focus.GleanMetrics.Browser
+import org.mozilla.focus.GleanMetrics.LegacyIds
+import org.mozilla.focus.GleanMetrics.Pings
+import org.mozilla.focus.ext.components
+import org.mozilla.focus.telemetry.TelemetryWrapper.isTelemetryEnabled
+import org.mozilla.focus.utils.Settings
+import java.util.UUID
+
+class GleanMetricsService(private val context: Context) : MetricsService {
+
+    private val activationPing = ActivationPing(context)
+
+    /**
+     * Initialization flag to
+     */
+    private var initialized = false
+
+    override fun initialize(context: Context) {
+        if (initialized) {
+            return
+        }
+        initialized = true
+
+        val components = context.components
+        val settings = Settings.getInstance(context)
+        val telemetryEnabled = isTelemetryEnabled(context)
+
+        Glean.initialize(
+            applicationContext = context,
+            uploadEnabled = telemetryEnabled,
+            configuration = Configuration(
+                channel = BuildConfig.FLAVOR,
+                httpClient = ConceptFetchHttpUploader(
+                    client = lazy(LazyThreadSafetyMode.NONE) { components.client },
+                    usePrivateRequest = true
+                )
+            ),
+            buildInfo = GleanBuildInfo.buildInfo
+        )
+
+        Glean.registerPings(Pings)
+
+        // Do this immediately after init.
+        MainScope().launch {
+
+            // Wait for preferences to be collected before we send the activation ping.
+            collectPrefMetrics(components, settings).await()
+
+            // Set the client ID in Glean as part of the deletion-request.
+            LegacyIds.clientId.set(UUID.fromString(TelemetryWrapper.clientId))
+
+            components.store.waitForSelectedOrDefaultSearchEngine { searchEngine ->
+                if (searchEngine != null) {
+                    Browser.defaultSearchEngine.set(getDefaultSearchEngineIdentifierForTelemetry(context))
+                }
+
+                activationPing.checkAndSend()
+            }
+        }
+    }
+
+    override fun track(event: Event) {
```

Comment:
I'd recommend not following this approach :-)

Reply:
What would be an alternative? I'm open to other options, although I prefer not to have Glean generated code around the code base.

It tends to be a problem if there are build issues with the parser and code generation fails (it those cases it's easier to disable one file that contains all the generated code), mocking out these classes in tests add overhead.

Another reason that is not in our direct path, but worth noting, is that forks of our browsers tend to remove our telemetry in place of their own, so having that in one place would also make that easier.

Reply:
> 
> 
> What would be an alternative? I'm open to other options, although I prefer not to have Glean generated code around the code base.

The alternative is the recommended path: calling the Glean APIs where needed, instead of dispatching.
Note that the tone of this is, again, of a recommendation (albeit I _strongly recommend_ doing that): your team owns Focus so it's ultimately your decision. However, the current approach has a series of documented shortcomings that bit us already in Fenix:

- [A perf investigation](https://docs.google.com/document/d/1GlqkaAKMcij0RLPs9aWyPmscR8eXgRn_xxy8JSc0CdU/edit) found that `.track` is a performance problem.
- [A Fenix + Glean postmortem](https://docs.google.com/document/d/198L01tgQYq10Th9l3cIeJd-Gk_iISGmGnow5_Ubw02w/edit) identified that this abstraction made it unsustainable to add new metrics at the pace required by Fenix.
- [An investigation by our team](https://bugzilla.mozilla.org/show_bug.cgi?id=1572160) with the help of some Fenix folks to identify a better option to streamline metric addition..

> It tends to be a problem if there are build issues with the parser and code generation fails (it those cases it's easier to disable one file that contains all the generated code), mocking out these classes in tests add overhead.

Do you have a specific example in mind? If code generation fails, that's a bug that should be fixed by us rather than mocked/ignored. Moreover, Glean is pretty mature now and we don't expect these to happen frequently (they did not happen very frequently even early on FWIW :-) ).

> Another reason that is not in our direct path, but worth noting, is that forks of our browsers tend to remove our telemetry in place of their own, so having that in one place would also make that easier.

While I can sympathise with this, we already perform our due diligence by making disabling data collection extremely simple. Folks who want to maintain a fork can still do so, committing a little time stripping Glean if they wish so.

Reply:
Thanks for detailed response that gives really good context!

I want to move forward with this patch, so I'll take your recommendation on that and if we can discuss more in the future if things need to change. :)

Reply:
> 
> 
> Thanks for detailed response that gives really good context!

My pleasure! It took a bit of bugzilla-archeology but now I have the context back as well :-)

> I want to move forward with this patch, so I'll take your recommendation on that and if we can discuss more in the future if things need to change. :)

Thank you so much! We're more than happy to discuss changes going forward and help with any problem that might arise!

Reply:
> My pleasure! It took a bit of bugzilla-archeology but now I have the context back as well :-)

Will reference your comment in future discussions so the effort doesn't go to waste! 😄